core/src/main/java/com/linecorp/armeria/server/cors/CorsPolicy.java

```diff
@@ -0,0 +1,285 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.cors;
+
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import javax.annotation.Nullable;
+
+import com.google.common.base.Joiner;
+import com.google.common.base.MoreObjects;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+
+import com.linecorp.armeria.common.DefaultHttpHeaders;
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.server.cors.CorsConfig.ConstantValueSupplier;
+import com.linecorp.armeria.server.cors.CorsConfig.InstantValueSupplier;
+
+import io.netty.util.AsciiString;
+
+public final class CorsPolicy {
+
+    private static final String DELIMITER = ",";
+    private static final Joiner HEADER_JOINER = Joiner.on(DELIMITER);
+    private final Set<String> origins;
+    private final boolean credentialsAllowed;
+    private final boolean nullOriginAllowed;
+    private final long maxAge;
+    private final Set<AsciiString> exposedHeaders;
+    private final Set<HttpMethod> allowedRequestMethods;
+    private final Set<AsciiString> allowedRequestHeaders;
+    private final Map<AsciiString, Supplier<?>> preflightResponseHeaders;
+
+    CorsPolicy(final Set<String> origins, boolean credentialsAllowed, long maxAge,
+               boolean nullOriginAllowed, Set<AsciiString> exposedHeaders,
+               Set<AsciiString> allowedRequestHeaders, Set<HttpMethod> allowedRequestMethods,
+               boolean preflightResponseHeadersDisabled,
+               Map<AsciiString, Supplier<?>> preflightResponseHeaders) {
+        this.origins = origins;
+        this.credentialsAllowed = credentialsAllowed;
+        this.maxAge = maxAge;
+        this.nullOriginAllowed = nullOriginAllowed;
+        this.exposedHeaders = ImmutableSet.copyOf(exposedHeaders);
+        this.allowedRequestMethods = EnumSet.copyOf(allowedRequestMethods);
+        this.allowedRequestHeaders = ImmutableSet.copyOf(allowedRequestHeaders);
+
+        if (preflightResponseHeadersDisabled) {
+            this.preflightResponseHeaders = Collections.emptyMap();
+        } else if (preflightResponseHeaders.isEmpty()) {
+            this.preflightResponseHeaders = ImmutableMap.of(
+                    HttpHeaderNames.DATE, InstantValueSupplier.INSTANCE,
+                    HttpHeaderNames.CONTENT_LENGTH, ConstantValueSupplier.ZERO);
+        } else {
+            this.preflightResponseHeaders = ImmutableMap.copyOf(preflightResponseHeaders);
+        }
+    }
+
+    /**
+     * Returns the allowed origin. This can either be a wildcard or an origin value.
```

Comment:
Could you add:

  This method returns the first specified origin if this policy has more than one origin.